File: Robotics_Fall_2018/ros-pkg-samples/self_drive/src/self_drive.cpp
#include <ros/ros.h>
#include <sensor_msgs/LaserScan.h>
#include <geometry_msgs/Twist.h>
#include <functional>
#include "self_drive/tb3driver.h"

TB3Driver* tb3driver;

void ldsCallback(const sensor_msgs::LaserScan::ConstPtr& scan)
{
    ROS_INFO("receive range size = %d, first = %.3f",
             (int)scan->ranges.size(), scan->ranges[0]);
    ROS_INFO("receive inten size = %d, first = %.3f",
             (int)scan->intensities.size(), scan->intensities[0]);
    tb3driver->ldsCallback(scan);
}

int main(int argc, char **argv)
{
	// 노드 네임 초기화
	ros::init(argc, argv, "self_drive");
	// Node handle declaration for communication with ROS system
	ros::NodeHandle nh;
	
	// create publisher: 
    ros::Publisher velopub = nh.advertise<geometry_msgs::Twist>("cmd_vel", 100);
	// set publishing frequency
	ros::Rate loop_rate(10);

    // 반드시 subscriber가 돌아가기 전에 생성
    tb3driver = new TB3Driver(&velopub);
	
	// create subscriber
    ros::Subscriber ldssub = nh.subscribe("scan", 1, ldsCallback);

    while(ros::ok())
    {
        tb3driver->publishVelocity();
        ros::spinOnce();
        loop_rate.sleep();
    }
	return 0;
}

File: Robotics_Fall_2019/ros_tutorial/self_drive/src/self_drive.py
#!/path/to/venv/bin/python

import rospy
from geometry_msgs.msg import Twist
from sensor_msgs.msg import LaserScan

class SelfDrive:
    def __init__(publisher):
        self.publisher = publisher

    def lds_callback(scan):
        # scan 분석 후 속도 결정
        # ...
        turtle_vel = Twist()
        # 전진 속도 및 회전 속도 지정
        turtle_vel.linear.x = 0.1
        turtle_vel.angular.z = 0.2
        # 속도 출력
        publisher.publish(turtle_vel)

rospy.init_node('self_drive')
publisher = rospy.Publisher('cmd_vel', TimePose, queue_size=1)
driver = SelfDrive(publisher)
subscriber = rospy.Subscriber('scan', LaserScan, 
                              lambda scan: driver.lds_callback(scan))
rospy.spin()

